Mathlib/Algebra/Lie/Basic.lean

```diff
@@ -0,0 +1,1134 @@
+  coe_injective <| funext h
+#align lie_equiv.ext LieEquiv.ext
+
+instance : One (L₁ ≃ₗ⁅R⁆ L₁) :=
+  ⟨{ (1 : L₁ ≃ₗ[R] L₁) with map_lie' := rfl }⟩
+
+@[simp]
+theorem one_apply (x : L₁) : (1 : L₁ ≃ₗ⁅R⁆ L₁) x = x :=
+  rfl
+#align lie_equiv.one_apply LieEquiv.one_apply
+
+instance : Inhabited (L₁ ≃ₗ⁅R⁆ L₁) :=
+  ⟨1⟩
+
+/-- Lie algebra equivalences are reflexive. -/
+def refl : L₁ ≃ₗ⁅R⁆ L₁ :=
+  1
+#align lie_equiv.refl LieEquiv.refl
+
+@[simp]
+theorem refl_apply (x : L₁) : (refl : L₁ ≃ₗ⁅R⁆ L₁) x = x :=
+  rfl
+#align lie_equiv.refl_apply LieEquiv.refl_apply
+
+/-- Lie algebra equivalences are symmetric. -/
+@[symm]
+def symm (e : L₁ ≃ₗ⁅R⁆ L₂) : L₂ ≃ₗ⁅R⁆ L₁ :=
+  { LieHom.inverse e.toLieHom e.invFun e.left_inv e.right_inv, e.toLinearEquiv.symm with }
+#align lie_equiv.symm LieEquiv.symm
+
+@[simp]
+theorem symm_symm (e : L₁ ≃ₗ⁅R⁆ L₂) : e.symm.symm = e := by
+  ext
+  rfl
+#align lie_equiv.symm_symm LieEquiv.symm_symm
+
+@[simp]
+theorem apply_symm_apply (e : L₁ ≃ₗ⁅R⁆ L₂) : ∀ x, e (e.symm x) = x :=
+  e.toLinearEquiv.apply_symm_apply
+#align lie_equiv.apply_symm_apply LieEquiv.apply_symm_apply
+
+@[simp]
+theorem symm_apply_apply (e : L₁ ≃ₗ⁅R⁆ L₂) : ∀ x, e.symm (e x) = x :=
+  e.toLinearEquiv.symm_apply_apply
+#align lie_equiv.symm_apply_apply LieEquiv.symm_apply_apply
+
+@[simp]
+theorem refl_symm : (refl : L₁ ≃ₗ⁅R⁆ L₁).symm = refl :=
+  rfl
+#align lie_equiv.refl_symm LieEquiv.refl_symm
+
+/-- Lie algebra equivalences are transitive. -/
+@[trans]
+def trans (e₁ : L₁ ≃ₗ⁅R⁆ L₂) (e₂ : L₂ ≃ₗ⁅R⁆ L₃) : L₁ ≃ₗ⁅R⁆ L₃ :=
+  { LieHom.comp e₂.toLieHom e₁.toLieHom, LinearEquiv.trans e₁.toLinearEquiv e₂.toLinearEquiv with }
+#align lie_equiv.trans LieEquiv.trans
+
+@[simp]
+theorem self_trans_symm (e : L₁ ≃ₗ⁅R⁆ L₂) : e.trans e.symm = refl :=
+  ext e.symm_apply_apply
+#align lie_equiv.self_trans_symm LieEquiv.self_trans_symm
+
+@[simp]
+theorem symm_trans_self (e : L₁ ≃ₗ⁅R⁆ L₂) : e.symm.trans e = refl :=
+  e.symm.self_trans_symm
+#align lie_equiv.symm_trans_self LieEquiv.symm_trans_self
+
+@[simp]
+theorem trans_apply (e₁ : L₁ ≃ₗ⁅R⁆ L₂) (e₂ : L₂ ≃ₗ⁅R⁆ L₃) (x : L₁) : (e₁.trans e₂) x = e₂ (e₁ x) :=
+  rfl
+#align lie_equiv.trans_apply LieEquiv.trans_apply
+
+@[simp]
+theorem symm_trans (e₁ : L₁ ≃ₗ⁅R⁆ L₂) (e₂ : L₂ ≃ₗ⁅R⁆ L₃) :
+    (e₁.trans e₂).symm = e₂.symm.trans e₁.symm :=
+  rfl
+#align lie_equiv.symm_trans LieEquiv.symm_trans
+
+protected theorem bijective (e : L₁ ≃ₗ⁅R⁆ L₂) : Function.Bijective ((e : L₁ →ₗ⁅R⁆ L₂) : L₁ → L₂) :=
+  e.toLinearEquiv.bijective
+#align lie_equiv.bijective LieEquiv.bijective
+
+protected theorem injective (e : L₁ ≃ₗ⁅R⁆ L₂) : Function.Injective ((e : L₁ →ₗ⁅R⁆ L₂) : L₁ → L₂) :=
+  e.toLinearEquiv.injective
+#align lie_equiv.injective LieEquiv.injective
+
+protected theorem surjective (e : L₁ ≃ₗ⁅R⁆ L₂) :
+    Function.Surjective ((e : L₁ →ₗ⁅R⁆ L₂) : L₁ → L₂) :=
+  e.toLinearEquiv.surjective
+#align lie_equiv.surjective LieEquiv.surjective
+
+/-- A bijective morphism of Lie algebras yields an equivalence of Lie algebras. -/
+@[simps!]
```

Comment:
@LukasMias How does `simps!` differ from `simps`? Why is it preferred here?

Reply:
`simps!` also works for "projections" that are not literally structure projections. At least that's my primitive understanding. And so with new structures we need this more often.